rasa/core/story_conflict.py

```diff
@@ -0,0 +1,252 @@
+from typing import List, Optional, Dict, Text
+
+from rasa.core.actions.action import ACTION_LISTEN_NAME
+from rasa.core.domain import PREV_PREFIX, Domain
+from rasa.core.events import ActionExecuted, Event
+from rasa.core.featurizers import MaxHistoryTrackerFeaturizer
+from rasa.nlu.constants import MESSAGE_INTENT_ATTRIBUTE
+from rasa.core.training.generator import TrackerWithCachedStates
+
+
+class StoryConflict:
+    def __init__(
+        self, sliced_states: List[Optional[Dict[Text, float]]],
+    ):
+        self.sliced_states = sliced_states
+        self.hash = hash(str(list(sliced_states)))
+        self._conflicting_actions = {}  # {"action": ["story_1", ...], ...}
+        self.correct_response = None
+
+    @staticmethod
+    def find_conflicts(
+        trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+    ) -> List:
+        """
+        Generate a list of StoryConflict objects, describing
+        conflicts in the given trackers.
+        :param trackers: Trackers in which to search for conflicts
+        :param domain: The domain
+        :param max_history: The maximum history length to be
+        taken into account
+        :return: List of conflicts
+        """
+
+        # We do this in two steps, to reduce memory consumption:
+
+        # Create a 'state -> list of actions' dict, where the state is
+        # represented by its hash
+        rules = StoryConflict._find_conflicting_states(trackers, domain, max_history)
+
+        # Iterate once more over all states and note the (unhashed) state,
+        # for which a conflict occurs
+        conflicts = StoryConflict._build_conflicts_from_states(
```

Comment:
Does it actually save a lot of memory? If we would it in one step, we would store another string `story_name` for each action right? That doesn't seem to bad?

Reply:
No, you need to store the un-hashed sliced state, so you can print out what came before the conflicting action.

Reply:
This is also in anticipation of printing more than one prior action for semi-automatic conflict resolution.